### PR TITLE
[WIP] Streaming parser bufferedsink

### PR DESCRIPTION
**Pull request by WyriHaximus**
Just like the bufferedsink from stream and #70 but this sink takes a streaming body parser and buffers any post field, body data, or uploaded file. Once the full request it in the promise will resolve.

Depends on #69 and #62 
- [X] BufferedSink for streaming parsers
- [ ] Class/Method Documentation
- [ ] Readme Documentation
- [ ] Split up sink in a file, post, and body sink
- [ ] Merge master in when #69 and #62 are in
- [ ] Ensure tests are up to date and passing
- [ ] Squash on merge

### CODE REVIEW

**Comment by jsor** [comment]
What about splitting this into 3 distinct sinks: a body, post and files sink. The sinks could be composed with parts you need. The resolution array from the current implementation could be achieved with

```php
$promises = [
    'body'  => BodyBufferedSink::createPromise($parser),
    'post'  => PostBufferedSink::createPromise($parser),
    'files' => FilesBufferedSink::createPromise($parser),
];
React\Promise\all($promises)->done($result) {
   // $result contains 'body', 'post' and 'files `entries
});
```

**Comment by WyriHaximus** [comment]
That is actually a great idea :+1: !

**Comment by WyriHaximus** [comment]
@jsor updated the PR with the split up `BufferedSink`, left the `BufferedSink` in for reference for now.

File: tests/StreamingBodyParser/PostBufferedSinkTest.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+namespace React\Tests\Http\StreamingBodyParser;
+
+use Clue\React\Block;
+use React\EventLoop\Factory;
+use React\Http\StreamingBodyParser\PostBufferedSink;
+use React\Http\StreamingBodyParser\NoBodyParser;
+use React\Http\Request;
+use React\Tests\Http\TestCase;
+
+class PostPostBufferedSinkTest extends TestCase
```
Comment (jsor) [inline tests/StreamingBodyParser/PostBufferedSinkTest.php]: s/PostPost/Post

**Comment by jsor** [comment]
Maybe the `BufferedSink` can stay as

```php
class BufferedSink
{
    /**
     * @param ParserInterface $parser
     * @return PromiseInterface
     */
    public static function createPromise(ParserInterface $parser)
    {
        $promises = [
            'body'  => BodyBufferedSink::createPromise($parser),
            'post'  => PostBufferedSink::createPromise($parser),
            'files' => FilesBufferedSink::createPromise($parser),
        ];

        return Promise\all($promises);
    }
}
```

**Pull request closed without merging**